### ISSUE TITLE
.node-version file format very restricted

### ISSUE THREAD

**Issue by shadowspawn** [teaxyz/cli#539]
From my testing, it appears the `.node-version` file can not have a line-ending. I suggest at least allowing a unix line ending.

Disclaimer: I am not a tea user, but am interested in `.node-version` support across utilities:
- https://github.com/shadowspawn/node-version-usage

I struggled with the `cd` magic silently ignoring unexpected file formats, but discovered I do get error messages from `tea -SE` and made more progress with that. Here is trying unix line ending, windows line ending, a leading v, and then (hurrah!) a plain version number with no line ending. Running on Apple Silicon Mac.

```
% printf "19.7.0/n" > .node-version
% tea -SE                          
pantries sync’d ⎷

panic: spilt tea. we’re sorry and we’ll fix it… but you have to report the bug!

    https://github.com/teaxyz/cli/issues/new?title=panic%3Acouldn%E2%80%99t+parse%3A+.node-version

----------------------------------------------------->> attachment begin
Error: couldn’t parse: .node-version
    at supp (file:///private/var/folders/y3/yr3bxsy95c3gfwwj61ytbp000000gq/T/cbf93776/src/hooks/useVirtualEnv.ts:135:36)
    at async default (file:///private/var/folders/y3/yr3bxsy95c3gfwwj61ytbp000000gq/T/cbf93776/src/hooks/useVirtualEnv.ts:35:17)
    at async run (file:///private/var/folders/y3/yr3bxsy95c3gfwwj61ytbp000000gq/T/cbf93776/src/app.main.ts:31:30)
    at async file:///private/var/folders/y3/yr3bxsy95c3gfwwj61ytbp000000gq/T/cbf93776/src/app.ts:11:5
<<----------------------------------------------------- attachment end

% printf "19.7.0/r/n" > .node-version
% tea -SE                            
pantries sync’d ⎷

panic: spilt tea. we’re sorry and we’ll fix it… but you have to report the bug!

    https://github.com/teaxyz/cli/issues/new?title=panic%3Acouldn%E2%80%99t+parse%3A+.node-version

----------------------------------------------------->> attachment begin
Error: couldn’t parse: .node-version
    at supp (file:///private/var/folders/y3/yr3bxsy95c3gfwwj61ytbp000000gq/T/cbf93776/src/hooks/useVirtualEnv.ts:135:36)
    at async default (file:///private/var/folders/y3/yr3bxsy95c3gfwwj61ytbp000000gq/T/cbf93776/src/hooks/useVirtualEnv.ts:35:17)
    at async run (file:///private/var/folders/y3/yr3bxsy95c3gfwwj61ytbp000000gq/T/cbf93776/src/app.main.ts:31:30)
    at async file:///private/var/folders/y3/yr3bxsy95c3gfwwj61ytbp000000gq/T/cbf93776/src/app.ts:11:5
<<----------------------------------------------------- attachment end

% printf "v19.7.0" > .node-version
% tea -SE                         
pantries sync’d ⎷

panic: spilt tea. we’re sorry and we’ll fix it… but you have to report the bug!

    https://github.com/teaxyz/cli/issues/new?title=panic%3Acouldn%E2%80%99t+parse%3A+.node-version

----------------------------------------------------->> attachment begin
Error: couldn’t parse: .node-version
    at supp (file:///private/var/folders/y3/yr3bxsy95c3gfwwj61ytbp000000gq/T/cbf93776/src/hooks/useVirtualEnv.ts:135:36)
    at async default (file:///private/var/folders/y3/yr3bxsy95c3gfwwj61ytbp000000gq/T/cbf93776/src/hooks/useVirtualEnv.ts:35:17)
    at async run (file:///private/var/folders/y3/yr3bxsy95c3gfwwj61ytbp000000gq/T/cbf93776/src/app.main.ts:31:30)
    at async file:///private/var/folders/y3/yr3bxsy95c3gfwwj61ytbp000000gq/T/cbf93776/src/app.ts:11:5
<<----------------------------------------------------- attachment end

% printf "19.7.0" > .node-version 
% tea -SE                        
pantries sync’d ⎷
installed: ~/.tea/nodejs.org/v19.7.0
SRCROOT=/Users/tea/fixtures/dev
TEA_FILES=/Users/tea/fixtures/dev/.node-version
SSL_CERT_FILE=/Users/tea/.tea/curl.se/ca-certs/v2023.1.10/ssl/cert.pem
GIT_SSL_CAINFO=/Users/tea/.tea/curl.se/ca-certs/v2023.1.10/ssl/cert.pem
PATH=/Users/tea/.tea/unicode.org/v71.1.0/bin:/Users/tea/.tea/unicode.org/v71.1.0/sbin:/Users/tea/.tea/openssl.org/v1.1.1t/bin:/Users/tea/.tea/nodejs.org/v19.7.0/bin:/Users/tea/.tea/tea.xyz/v*/bin:/Users/tea/.local/bin:/opt/homebrew/bin:/opt/homebrew/sbin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin
MANPATH=/Users/tea/.tea/unicode.org/v71.1.0/share/man:/Users/tea/.tea/zlib.net/v1.2.13/share/man:/Users/tea/.tea/nodejs.org/v19.7.0/share/man:/usr/share/man:/opt/homebrew/share/man
PKG_CONFIG_PATH=/Users/tea/.tea/unicode.org/v71.1.0/lib/pkgconfig:/Users/tea/.tea/openssl.org/v1.1.1t/lib/pkgconfig:/Users/tea/.tea/zlib.net/v1.2.13/lib/pkgconfig
LIBRARY_PATH=/Users/tea/.tea/unicode.org/v71.1.0/lib:/Users/tea/.tea/openssl.org/v1.1.1t/lib:/Users/tea/.tea/zlib.net/v1.2.13/lib
LD_LIBRARY_PATH=/Users/tea/.tea/unicode.org/v71.1.0/lib:/Users/tea/.tea/openssl.org/v1.1.1t/lib:/Users/tea/.tea/zlib.net/v1.2.13/lib
CPATH=/Users/tea/.tea/unicode.org/v71.1.0/include:/Users/tea/.tea/openssl.org/v1.1.1t/include:/Users/tea/.tea/zlib.net/v1.2.13/include:/Users/tea/.tea/nodejs.org/v19.7.0/include
XDG_DATA_DIRS=/Users/tea/.tea/unicode.org/v71.1.0/share:/Users/tea/.tea/zlib.net/v1.2.13/share:/Users/tea/.tea/nodejs.org/v19.7.0/share
DYLD_FALLBACK_LIBRARY_PATH=/Users/tea/.tea/unicode.org/v71.1.0/lib:/Users/tea/.tea/openssl.org/v1.1.1t/lib:/Users/tea/.tea/zlib.net/v1.2.13/lib
LDFLAGS=-Wl,-rpath,/Users/tea/.tea
TEA_PREFIX=/Users/tea/.tea
```

**Comment by jhheider** [comment]
Generally, this is a bug because `tea` parses by lines, and your second line, `""` confuses it. However, it doesn't seem like this should be the case:
```typescript
    if (_if(".node-version")) {
      const constraint = semver.Range.parse((await f!.read()).trim())
      if (!constraint) throw new Error('couldn’t parse: .node-version')
      pkgs.push({ project: "nodejs.org", constraint })
    }
```

With `tea 0.27.0`, I seem to get the right results:
```sh
$ echo -e "^20\n" >.node-version
$ tea -E
installed: ~/.tea/nodejs.org/v20.0.0
...
$ hexdump .node-version
0000000 325e 0a30 000a
0000005
tea --version
tea 0.27.0
```

`git log` suggests it's used `.read().trim()` since at least Feb 1. Is it possible your system doesn't accept forward slash for escape sequences; mine (macOS) does not:
```sh
$ printf "^20/n"
^20/n%                                                                                                                                                                   
$ printf "^20\n"
^20
```

So, more information needed. :) `tea` version, I think, and possibly shell, since we're running on the same hardware.

**Comment by shadowspawn** [comment]
Drat, I was doing hexdump at various points but not carefully enough. I will recheck, may be some user error involved!

**Comment by shadowspawn** [comment]
```
% tea --version
tea 0.28.1
% echo $SHELL
/bin/zsh
```

Ok, the line-ending issues were my mistake! (My apologies.)

I do see some interesting behaviour when there is a leading `v`. Not sure if you care about a leading `v`! Feel free to close this issue, and/or I could open a new issue if you want to track that with a cleaner issue than this one.

On first sync it does appear to install the node version but then it does not activate. A subsequent sync shows an error. 

```
% rm -rf ~/.tea/nodejs.org           
% printf "v19.7.0\n" > .node-version
% hexdump -C .node-version          
00000000  76 31 39 2e 37 2e 30 0a                           |v19.7.0.|
00000008
% tea -SE                           
pantries sync’d ⎷
installed: ~/.tea/nodejs.org/v19.7.0
SRCROOT=/Users/tea/fixtures/dev
TEA_FILES=/Users/tea/fixtures/dev/.node-version
SSL_CERT_FILE=/Users/tea/.tea/curl.se/ca-certs/v2023.1.10/ssl/cert.pem
GIT_SSL_CAINFO=/Users/tea/.tea/curl.se/ca-certs/v2023.1.10/ssl/cert.pem
PATH=/Users/tea/.tea/unicode.org/v71.1.0/bin:/Users/tea/.tea/unicode.org/v71.1.0/sbin:/Users/tea/.tea/openssl.org/v1.1.1t/bin:/Users/tea/.tea/nodejs.org/v19.7.0/bin:/Users/tea/.tea/tea.xyz/v*/bin:/Users/tea/.local/bin:/opt/homebrew/bin:/opt/homebrew/sbin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin
MANPATH=/Users/tea/.tea/unicode.org/v71.1.0/share/man:/Users/tea/.tea/zlib.net/v1.2.13/share/man:/Users/tea/.tea/nodejs.org/v19.7.0/share/man:/usr/share/man
PKG_CONFIG_PATH=/Users/tea/.tea/unicode.org/v71.1.0/lib/pkgconfig:/Users/tea/.tea/openssl.org/v1.1.1t/lib/pkgconfig:/Users/tea/.tea/zlib.net/v1.2.13/lib/pkgconfig
LIBRARY_PATH=/Users/tea/.tea/unicode.org/v71.1.0/lib:/Users/tea/.tea/openssl.org/v1.1.1t/lib:/Users/tea/.tea/zlib.net/v1.2.13/lib
LD_LIBRARY_PATH=/Users/tea/.tea/unicode.org/v71.1.0/lib:/Users/tea/.tea/openssl.org/v1.1.1t/lib:/Users/tea/.tea/zlib.net/v1.2.13/lib
CPATH=/Users/tea/.tea/unicode.org/v71.1.0/include:/Users/tea/.tea/openssl.org/v1.1.1t/include:/Users/tea/.tea/zlib.net/v1.2.13/include:/Users/tea/.tea/nodejs.org/v19.7.0/include
XDG_DATA_DIRS=/Users/tea/.tea/unicode.org/v71.1.0/share:/Users/tea/.tea/zlib.net/v1.2.13/share:/Users/tea/.tea/nodejs.org/v19.7.0/share
DYLD_FALLBACK_LIBRARY_PATH=/Users/tea/.tea/unicode.org/v71.1.0/lib:/Users/tea/.tea/openssl.org/v1.1.1t/lib:/Users/tea/.tea/zlib.net/v1.2.13/lib
LDFLAGS=-Wl,-rpath,/Users/tea/.tea
TEA_PREFIX=/Users/tea/.tea
% cd .
% node --version
v16.20.0
% tea -SE
pantries sync’d ⎷

panic: spilt tea. we’re sorry and we’ll fix it… but you have to report the bug!

    https://github.com/teaxyz/cli/issues/new?title=panic%3Acouldn%E2%80%99t+parse%3A+.node-version

----------------------------------------------------->> attachment begin
Error: couldn’t parse: .node-version
    at supp (file:///private/var/folders/y3/yr3bxsy95c3gfwwj61ytbp000000gq/T/cbf93776/src/hooks/useVirtualEnv.ts:135:36)
    at async default (file:///private/var/folders/y3/yr3bxsy95c3gfwwj61ytbp000000gq/T/cbf93776/src/hooks/useVirtualEnv.ts:35:17)
    at async run (file:///private/var/folders/y3/yr3bxsy95c3gfwwj61ytbp000000gq/T/cbf93776/src/app.main.ts:31:30)
    at async file:///private/var/folders/y3/yr3bxsy95c3gfwwj61ytbp000000gq/T/cbf93776/src/app.ts:11:5
<<----------------------------------------------------- attachment end
```

**Comment by mxcl** [comment]
Can fix the leading v easy enough.

**Comment by mxcl** [comment]
How does it work? Does `v16` mean `^16` or `=16.0.0`?

**Comment by mxcl** [comment]
k using your doc for reference https://github.com/shadowspawn/node-version-usage thanks!

**Comment by mxcl** [comment]
k once #541 is merged I believe we comply as your document suggests.

**Comment by shadowspawn** [comment]
Fast work! Thanks.
